src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -48,60 +49,99 @@ internal static SafeChannelBindingHandle QueryChannelBinding(SafeSslHandle conte
 
         internal static SafeSslHandle AllocateSslContext(SslProtocols protocols, SafeX509Handle certHandle, SafeEvpPKeyHandle certKeyHandle, EncryptionPolicy policy, bool isServer, bool remoteCertRequired)
         {
-            SafeSslHandle context = null;
+            bool hasCertReference = false, 
+                hasCertKeyReference = false;
 
-            IntPtr method = GetSslMethod(protocols);
+            bool hasCertificateAndKey = 
+                certHandle != null && !certHandle.IsInvalid
+                && certKeyHandle != null && !certKeyHandle.IsInvalid;
 
-            using (SafeSslContextHandle innerContext = Ssl.SslCtxCreate(method))
+            try
             {
-                if (innerContext.IsInvalid)
+                if (hasCertificateAndKey)
                 {
-                    throw CreateSslException(SR.net_allocate_ssl_context_failed);
+                    certHandle.DangerousAddRef(ref hasCertReference);
+                    certKeyHandle.DangerousAddRef(ref hasCertKeyReference);
```

Comment:
Nah man, better to fail early and get before you start doing any unnecessary allocations. Why bother creating and having to clean up context when you don't know yet if the entire method body basically going to be a waste of time.